lib/libc/string/arch/x86/64/memcpy.c

```diff
@@ -0,0 +1,172 @@
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include <string.h>
+#include <sys/types.h>
+
+#define LONG_SIZE sizeof(long)
+#define LONG_MASK (LONG_SIZE - 1)
+#define BYTE_SIZE 1
+
+#ifndef TRUE
+#define TRUE 1
+#endif
+
+#ifndef FALSE
+#define FALSE 0
+#endif
+
+static inline void memcpy_inc(char *dest, const char *src, size_t len)
+{
+    __asm__ __volatile__ (
+            "cld        \n\t"
+            "rep movsb  \n\t"
+            :: "D" (dest), "S" (src), "c" (len)
+            );
+}
+
+static inline void memcpy64_inc(char *dest, const char *src, size_t len)
+{
+    __asm__ __volatile__ (
+            "cld        \n\t"
+            "rep movsq  \n\t"
+            :: "D" (dest), "S" (src), "c" (len >> 3)
+            );
+}
+
+static inline void memcpy_dec(char *dest, const char *src, size_t len)
+{
+    __asm__ __volatile__ (
+            "std        \n\t"
+            "rep movsb  \n\t"
+            "cld        \n\t"
+            :: "D" (dest + len - BYTE_SIZE),
+            "S" (src + len - BYTE_SIZE),
+            "c" (len)
+            );
+}
+
+static inline void memcpy64_dec(char *dest, const char *src, size_t len)
+{
+    __asm__ __volatile__ (
+            "std        \n\t"
+            "rep movsq  \n\t"
+            "cld        \n\t"
+            :: "D" (dest + len - LONG_SIZE),
+            "S" (src + len - LONG_SIZE),
+            "c" (len >> 3)
+            );
+}
+
+static inline int para_available(char *d, const char *s, size_t c)
+{
+    if ((NULL == d) || (NULL == s) || (0 == c) || (d == s))
+        return FALSE;
+
+    return TRUE;
+}
+
+static inline int para_with_different_offset(char *d, const char *s)
+{
+    if (((long)d ^ (long)s) & LONG_MASK)
+        return TRUE;
+
+    return FALSE;
+}
+
+static inline int para_count_too_small(size_t count)
+{
+    if (count < LONG_SIZE)
+        return TRUE;
+
+    return FALSE;
+}
+
+/*
+ * Two types of copy methods provided by Intel x86 architecutre:
+ * increment and decrement copy, DF flag controls copy direction.
+ * If source is larger than destination, increment copy selected,
+ * otherwise, decrement copy selected (since if increment copy used
+ * in this scenario, destination and source memory might overlapped
+ * for some region: src < dest < src + count, copy to destination may
+ * overwrite original source data, decrement copy sloves this issue).
+ *
+ * When decrement copy selected, processor copys data first, and then
+ * decreases RDI/RSI/RCX. In this case, we need to adjust RDI/RSI before
+ * copy to right address.
+ *
+ * More details about MOVS/MOVSB/MOVSW/MOVSD/MOVSQ please refer to
+ * <Intel 64 and IA-32 Architectures Software Developer's Manaual>
+ * Volume 2: Instruction Set Reference
+ */
+void *memcpy(void *dest, const void *src, size_t count)
```

Comment:
For the magenta branch of this we ended up just using rep movs directly: https://fuchsia.googlesource.com/magenta/+/master/kernel/lib/libc/string/arch/x86/